### ISSUE TITLE
Potential refactorings post v0.2

### ISSUE THREAD

**Issue by bobbinth** [0xPolygonMiden/air-script#150]
IR building process got pretty complicated and I think we can refactor it to make it more modular and easier to follow. I think the main thing is that there are quite a few functions which take a large number of arguments and keeping track of what's going on is difficult (at least for me). Below are a few suggestions. I'm not sure all of them will make sense as I haven't thought through everything in detail. So, maybe some (or all) of them will end up being discarded.

First, maybe we should separate building of constraint expression graph from the graph itself. Many method on `AlgebraicGraph` take a large number of arguments because they need access to the current context (e.g., `SymbolTable`, `VariableRoots`). Instead, maybe it would make sense to introduce a "builder" struct which would be responsible for maintaining the context and building the graph. Maybe something like this:
```Rust
struct ConstraintBuilder {
    symbol_table: SymbolTable,
    variable_roots: VariableRoots,
    constrained_boundaries: BTreeSet<ConstrainedBoundaries>,
    graph: AlgebraicGraph,
}
```

This struct would have methods like `insert_boundary_stmt()` and `insert_integrity_stmt()`. After all statements are inserted, it would be converted into a `Constraints` struct like so:
```Rust
impl ConstraintBuilder {
    fn into_constraints(self) -> Constraints
}
```
`ConstraintBuilder` would also contain most expression processing functions (e.g., `insertExpr` which is currently in `AlgebraicGraph`). Only the methods which modify the graph directly (e.g., `insert_op()`) would remain in `AlgebraicGraph`.

Second, we seem to be keeping track of some additional constraint info while inserting expressions into the graph. Specifically, we have this `ExprDetail` which is returned from many functions in the `AlgebraicGraph`. I wonder if we can take the same approach here as we did with computing constraint degree. Specifically, instead of keeping track of this additional info while we build a constraint, we could determine by recursively traversing the graph after the constraint is built. This would probably be less efficient, but I don't think we care about performance for this specific part.

Third, it may make sense to make some cosmetic changes to the symbol table. Specifically, we could adopt the nomenclature that symbol table stores symbols, not identifier types. I think this would make things a bit easier to understand as we'd use "symbol" everywhere rather than a mix of symbol, identifier type, element type etc.

**Comment by grjte** [comment]
I agree that a refactor would be good - I've been wanting to revisit this as well, so I'd like to work on this issue.

> First, maybe we should separate building of constraint expression graph from the graph itself. 

I think this makes sense.

> Second, we seem to be keeping track of some additional constraint info while inserting expressions into the graph. Specifically, we have this ExprDetail which is returned from many functions in the AlgebraicGraph. I wonder if we can take the same approach here as we did with computing constraint degree. Specifically, instead of keeping track of this additional info while we build a constraint, we could determine by recursively traversing the graph after the constraint is built. This would probably be less efficient, but I don't think we care about performance for this specific part.

I don't think I prefer this, since we do use this information right away, and I don't like the idea of essentially traveling the same path twice in a row every time we add a constraint to the graph. However, I do think we can tidy this up, and I think the approach to this will be related to your the `ConstraintBuilder` idea and adjustments to the symbol table.

I would hold off on this point until after we've addressed the other two points, then we can revisit it.

> Third, it may make sense to make some cosmetic changes to the symbol table. Specifically, we could adopt the nomenclature that symbol table stores symbols, not identifier types. I think this would make things a bit easier to understand as we'd use "symbol" everywhere rather than a mix of symbol, identifier type, element type etc.

I agree with this. Also, I'd like to introduce very basic scoping. This will help with adding the support for boundary variables, but it should also let us clean up some of the checks in the graph and move some errors into the symbol table that I think would be better there (for example, accessing a public input in a transition constraint, or a periodic column in a boundary constraint).

Finally, one thing that bothers me is that because of the way the types are currently working, we have a bunch of match statements that are nearly duplicated between the symbol table and the graph. I think this could be improved while making the transition to the `ConstraintBuilder` and adjusting the symbol table

**Comment by grjte** [comment]
This issue is being addressed in a few steps, some of which may change:

- [ ] restructure IR crate to introduce `ConstraintBuilder` and `Declarations` types (without moving graph builder functions out of the graph)
- [ ] simplify the symbol table validation of identifier accesses
- [ ] introduce `Symbol` type to the symbol table, with related renamings
- [ ] (maybe) introduce `Scope` to the symbol table to simplify scope handling and validation
- [ ] update the codebase to use the new `Symbol` type
- [ ] (maybe) simplify handling of `VariableRoots` to reduce code duplication between symbol table and graph-builder functions
- [ ] move graph-builder functions into the `ConstraintBuilder`

**Comment by tohrnii** [comment]
@grjte Can this issue be closed?

**Comment by tohrnii** [comment]
@grjte Can this issue be closed?

**Comment by grjte** [comment]
Closed by #190, #194, #195, #196, #197, #251, #252, #254, #255, #256, #258, #259
